Analyze dataframe and keep it until the next analysis (#3497)
freqtrade/freqtrade: docs/strategy-advanced.md

# Advanced Strategies

This page explains some advanced concepts available for strategies.
If you're just getting started, please be familiar with the methods described in the [Strategy Customization](strategy-customization.md) documentation and with the [Freqtrade basics](bot-basics.md) first.

> [Freqtrade basics](bot-basics.md) describes in which sequence each method defined below is called, which can be helpful to understand which method to use.

Reviewer: ```suggestion
[Freqtrade basics](bot-basics.md) describes in which sequence each method described below is called, which can be helpful to understand which method to use for your custom needs.
```

!!! Note

>     All callback methods described below should only be implemented in a strategy if they are also actively used.

Reviewer: Sorry, I did not understand the sentence (particularly, the usage of "also" here)...

Author: In short - dont blindly copy and paste it to the strategy

>     All callback methods described below should only be implemented in a strategy if they are also actively used.

Reviewer: ```suggestion
    All callback methods described below should only be implemented in a strategy if they are actually used.
```

            return True
        return False
```

## Bot loop start callback

> A simple callback which is called at the start of every bot iteration.
> This can be used to perform calculations which are pair independent.

Reviewer: ```suggestion
A simple callback which is called once at the start of every bot throttling iteration.
This can be used to perform calculations which are pair independent (apply to all pairs), loading of external data, etc.
```

``` python
import requests

class Awesomestrategy(IStrategy):

    # ... populate_* methods

    def bot_loop_start(self, **kwargs) -> None:
        """
        Called at the start of the bot iteration (one loop).
        Might be used to perform pair-independent tasks

>         (e.g. gather some remote ressource for comparison)

Reviewer: ```suggestion
        (e.g. gather some remote resource for comparison)
```

        :param **kwargs: Ensure to keep this here so updates to this won't break your strategy.
        """
        if self.config['runmode'].value in ('live', 'dry_run'):
            # Assign this to the class by using self.*
            # can then be used by populate_* methods
            self.remote_data = requests.get('https://some_remote_source.example.com')

```

## Bot order confirmation

### Trade entry (buy order) confirmation

> `confirm_trade_entry()` an be used to abort a trade entry at the latest second (maybe because the price is not what we expect).

Reviewer: ```suggestion
`confirm_trade_entry()` can be used to abort a trade entry at the latest second (maybe because the price is not what we expect).
```

                            time_in_force: str, **kwargs) -> bool:
        """
        Called right before placing a buy order.
        Timing for this function is critical, so avoid doing heavy computations or
        network requests in this method.

        For full documentation please go to https://www.freqtrade.io/en/latest/strategy-advanced/

        When not implemented by a strategy, returns True (always confirming).

        :param pair: Pair that's about to be bought.
        :param order_type: Order type (as configured in order_types). usually limit or market.
        :param amount: Amount in target (quote) currency that's going to be traded.
        :param rate: Rate that's going to be used when using limit orders
        :param time_in_force: Time in force. Defaults to GTC (Good-til-cancelled).
        :param **kwargs: Ensure to keep this here so updates to this won't break your strategy.
        :return bool: When True is returned, then the buy-order is placed on the exchange.
            False aborts the process
        """
        return True

```

### Trade exit (sell order) confirmation

> `confirm_trade_exit()` an be used to abort a trade exit (sell) at the latest second (maybe because the price is not what we expect).

Reviewer: ```suggestion
`confirm_trade_exit()` can be used to abort a trade exit (sell) at the latest second (maybe because the price is not what we expect).
```
